README.md
# ws-messaging

[![NPM Version](https://badge.fury.io/js/ws-messaging.svg)](https://badge.fury.io/js/ws-messaging)
[![Build Status](https://travis-ci.org/an-sh/ws-messaging.svg?branch=master)](https://travis-ci.org/an-sh/ws-messaging)
[![Appveyor status](https://ci.appveyor.com/api/projects/status/d14wp6ei50tmqy49/branch/master?svg=true)](https://ci.appveyor.com/project/an-sh/ws-messaging)
[![Coverage Status](https://codecov.io/gh/an-sh/ws-messaging/branch/master/graph/badge.svg)](https://codecov.io/gh/an-sh/ws-messaging)
[![Dependency Status](https://david-dm.org/an-sh/ws-messaging.svg)](https://david-dm.org/an-sh/ws-messaging)
[![JavaScript Style Guide](https://img.shields.io/badge/code%20style-standard-brightgreen.svg)](http://standardjs.com/)

Just a really thin abstraction layer on top of WebSocket for Node.js
and Web Browsers with Promise and EventEmitter based APIs.

### Features

- Send and receive notifications (events) via an EventEmitter API.

- Request-reply API using promises (works in both directions, without
  any connection blocking during processing).

- Built-in auth via WebSocket messages exchange (no more query
  strings).

- Auto reconnection is provided.

- Binary messages support via custom encoders/decoders.

- Reasonable client size: around 14KB minified, 5KB gziped.

## Table of Contents

- [Background](#background)
- [Installation](#installation)
- [Usage](#usage)
- [API](#api)
- [Network format description](#network-format-description)
- [Data validation](#data-validation)
- [Contribute](#contribute)
- [License](#license)


## Background

Read this
[article](https://medium.com/@an_sh_1/a-better-websocket-api-for-web-browsers-and-node-js-ws-messaging-6f7826242932)
for more background information.


## Installation

```sh
npm i ws-messaging
```

## Usage

On a server:

```javascript
const Server = require('ws-messaging')

const port = 8000

function connectionHook (client, authData) {
  // check an authData
  // then assign client events handlers
  // return a promise
}

let server = new Server({port}, {connectionHook})
```

On a client:

```javascript
const Client = require('ws-messaging/client')

const url = `ws://${HOST}:${PORT}`
const auth = { /* will be authData in connectionHook */ }

let client = new Client(url, {auth})

client.on('someEvent', (...data) => { /* do smth */ })

client.register('someMethod', (...args) => { /* do smth, return a promise */ })

client.on('connect', () => {
  /* now this client can send messages */
  client.send('myEvent', ...someData)
  /* or use request-reply (RPC) API */
  client.invoke('myMethod', ...someArgs)
    .then(result => { /* do smth */ })
    .catch(error => { /* do smth */ })
})
```

Essentially there are two usage patterns that are working in both
directions. Fire and forget via `send`/`on`, and RPC-style via
`invoke`/`register`. Unlike `on`, only a single handler function can
be registered per a method name.

## API

[Server API](https://an-sh.github.io/ws-messaging/2.0/Server.html) and
[Client API](https://an-sh.github.io/ws-messaging/2.0/Client.html)
documentation is available online.

## Network format description

This section describes what data is actually passed to an encoder.

There are only two types of messages. The first one is for normal
messages:

```javascript
{
  name: string,
  args: Array,
  id?: number
}
```

An `id` field is present for `invoke` calls. The second one is for
ack (replies for `invoke` calls) messages:

```javascript
{
  id: number
  result?: Object
  error?: Object
}
```

Either a `result` or an `error` field is included. Note that an
`error` is the value returned by an `errorFormatter`, by default
`String` is used as an `errorFormatter`.

## Data validation

All incoming data must be validated on a server side, including errors
that are passed to a catch callback. By default only the
[network format](#network-format-description) itself is
validated. Validation can be made by a custom decoder (useful when a
decoder is already using some scheme) or via a `receiveHook`, or
inside a handler itself (useful for registered procedures). When
validation is done inside `decoder`/`receiveHook`, just throw an error
or reject a promise to fail a validation and prevent handlers
execution. Also note that a promise returned by `invoke` can be
rejected locally either with `Client.ConnectionError` or with
`Client.TimeoutError`.

## Contribute

If you encounter a bug in this package, please submit a bug report to
github repo [issues](https://github.com/an-sh/ws-messaging/issues).

PRs are also accepted.

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/an-sh/ws-messaging/compare/v1.0.1...v2.0.0) (2019-12-14)


### ⚠ BREAKING CHANGES

* Now only Node.js version 10 and later are supported
* Remove browserify

* updates ([825b77c](https://github.com/an-sh/ws-messaging/commit/825b77c74393efab39ef0bda43a017d8bd0796a5))

<a name="1.0.1"></a>
## [1.0.1](https://github.com/an-sh/ws-messaging/compare/v1.0.0...v1.0.1) (2018-06-23)


### Bug Fixes

* silence not open warnings, closes [#3](https://github.com/an-sh/ws-messaging/issues/3) ([f985cfe](https://github.com/an-sh/ws-messaging/commit/f985cfe))



<a name="1.0.0"></a>
# [1.0.0](https://github.com/an-sh/ws-messaging/compare/v0.8.0...v1.0.0) (2018-01-05)


### Bug Fixes

* remove Promise polyfill ([f28cdd0](https://github.com/an-sh/ws-messaging/commit/f28cdd0))


### BREAKING CHANGES

* remove Promise polyfill



<a name="0.7.2"></a>
## [0.7.2](https://github.com/an-sh/ws-messaging/compare/v0.7.1...v0.7.2) (2017-01-19)


### Bug Fixes

* actually fix async receiveHook handling ([66934c9](https://github.com/an-sh/ws-messaging/commit/66934c9))



<a name="0.7.1"></a>
## [0.7.1](https://github.com/an-sh/ws-messaging/compare/v0.7.0...v0.7.1) (2017-01-17)


### Bug Fixes

* don't ignore promises returned by the receiveHook ([4547cbf](https://github.com/an-sh/ws-messaging/commit/4547cbf))



<a name="0.7.0"></a>
# [0.7.0](https://github.com/an-sh/ws-messaging/compare/v0.6.0...v0.7.0) (2017-01-06)


### Bug Fixes

* don't allow sending data when client is not connected ([64e3cb2](https://github.com/an-sh/ws-messaging/commit/64e3cb2))


### Code Refactoring

* start attempts counting from 1 ([9f6da34](https://github.com/an-sh/ws-messaging/commit/9f6da34))
* tweak default ping settings ([0455477](https://github.com/an-sh/ws-messaging/commit/0455477))


### Features

* add encodeMessage to Server ([800d081](https://github.com/an-sh/ws-messaging/commit/800d081))
* add sendHook ([b5fc8ea](https://github.com/an-sh/ws-messaging/commit/b5fc8ea))
* add support for sending already encoded messages ([6c6d065](https://github.com/an-sh/ws-messaging/commit/6c6d065))
* emitEncoded alias for sendEncoded ([5baf109](https://github.com/an-sh/ws-messaging/commit/5baf109))
* pass additional options to ws socket constructor, fixes [#1](https://github.com/an-sh/ws-messaging/issues/1) ([32237ad](https://github.com/an-sh/ws-messaging/commit/32237ad))


### BREAKING CHANGES

* Start attempts counting from 1.
* Tweak default ping settings.



<a name="0.6.0"></a>
# [0.6.0](https://github.com/an-sh/ws-messaging/compare/v0.5.3...v0.6.0) (2016-11-29)


### Bug Fixes

* transpiled code usage on node 4.x ([f267ce9](https://github.com/an-sh/ws-messaging/commit/f267ce9))


### Chores

* update dependencies ([f7b8e54](https://github.com/an-sh/ws-messaging/commit/f7b8e54))


### BREAKING CHANGES

* Drop node 0.12 support.



<a name="0.5.3"></a>
## [0.5.3](https://github.com/an-sh/ws-messaging/compare/v0.5.2...v0.5.3) (2016-10-27)



<a name="0.5.2"></a>
## [0.5.2](https://github.com/an-sh/ws-messaging/compare/v0.5.1...v0.5.2) (2016-10-27)


### Features

* add client socket constructor customisation ([b9b76e0](https://github.com/an-sh/ws-messaging/commit/b9b76e0))



<a name="0.5.1"></a>
## [0.5.1](https://github.com/an-sh/ws-messaging/compare/v0.5.0...v0.5.1) (2016-10-21)



<a name="0.5.0"></a>
# [0.5.0](https://github.com/an-sh/ws-messaging/compare/v0.4.1...v0.5.0) (2016-10-12)


### Bug Fixes

* ensure usage of formatter for all errors ([bc2dfad](https://github.com/an-sh/ws-messaging/commit/bc2dfad))
* make getClient to work inside onConnect ([1121efa](https://github.com/an-sh/ws-messaging/commit/1121efa))


### Code Refactoring

* rename parsing error ([ebfa845](https://github.com/an-sh/ws-messaging/commit/ebfa845))


### BREAKING CHANGES

* Rename ParsingError to preprocessingError.
* Error formatter is used on all errors, ParsingError and
RPCError wrappers have been removed.



<a name="0.4.1"></a>
## [0.4.1](https://github.com/an-sh/ws-messaging/compare/v0.4.0...v0.4.1) (2016-10-09)


### Bug Fixes

* closing reconnecting client ([82ee47c](https://github.com/an-sh/ws-messaging/commit/82ee47c))



<a name="0.4.0"></a>
# [0.4.0](https://github.com/an-sh/ws-messaging/compare/v0.3.0...v0.4.0) (2016-10-09)


### Features

* auto reconnect ([9a36be2](https://github.com/an-sh/ws-messaging/commit/9a36be2))
* handle auth timeout on a client-side ([7645801](https://github.com/an-sh/ws-messaging/commit/7645801))


### BREAKING CHANGES

* auto reconnect is enabled by default



<a name="0.3.0"></a>
# [0.3.0](https://github.com/an-sh/ws-messaging/compare/v0.2.0...v0.3.0) (2016-09-28)


### Bug Fixes

* returning an empty result from RPC ([103d0e9](https://github.com/an-sh/ws-messaging/commit/103d0e9))
* use standalone bundle ([c6fca77](https://github.com/an-sh/ws-messaging/commit/c6fca77))


### Code Refactoring

* allow reconnect on terminated clients ([ab0b806](https://github.com/an-sh/ws-messaging/commit/ab0b806))
* rename option ([2d95d5e](https://github.com/an-sh/ws-messaging/commit/2d95d5e))
* rename option ([b117cc4](https://github.com/an-sh/ws-messaging/commit/b117cc4))


### Features

* emit errors without throwing on 0 listeners ([7675db2](https://github.com/an-sh/ws-messaging/commit/7675db2))
* send pings for network failures detection ([2c7ee8d](https://github.com/an-sh/ws-messaging/commit/2c7ee8d))
* separate ping timeout option ([0ba3ef3](https://github.com/an-sh/ws-messaging/commit/0ba3ef3))
* use emit as an alias for send ([a142bed](https://github.com/an-sh/ws-messaging/commit/a142bed))


### BREAKING CHANGES

* Server option is renamed to WebSocketServer.
* terminated state now is not checked by the reconnect
method.
* rename ackWaitTimeout to ackTimeout.
* emit method semantics changed.




src/Server.js
// @flow
'use strict'

const Client = require('./Client')
const EventEmitter = require('eventemitter3')
const WebSocket = require('ws')
const WebSocketServer = require('ws').Server
const uid = require('uid-safe')
const { CLOSE_FORBIDDEN } = require('./constants')
const { assign, attempt, fromCallback } = require('./utils')

const defaults = {
  WebSocketServer,
  connectionHook: null,
  authTimeout: 20000
}

/**
 * Connection hook is run when a client connects to a server. The
 * result is used as an auth reply data. May also return promises for
 * an asynchronous execution. If the promise is rejected or an error
 * is thrown, then auth has failed and the socket will be closed.
 *
 * @callback Server.ConnectionHook
 *
 * @param {Client} client Client.
 * @param {Object} auth Auth data.
 * @return {Promise<Object|undefined>|Object|undefined} Auth reply
 * data.
 */
/* :: type ConnectionHook = (client: Client, data?: any) => Promise<any> */

/**
 * @typedef {Object} Server.ServerOptions
 *
 * @property {Server.ConnectionHook} [connectionHook] Connection
 * hook.
 * @property {Object} [WebSocketServer] Alternative constructor for ws
 * server.
 * @property {number} [authTimeout=20000] Auth message wait timeout in
 * ms.
 */
/* ::
type ServerOptions = { WebSocketServer?: constructor,
                       connectionHook?: ConnectionHook,
                       authTimeout?: number
                     }
*/

/* :: type SocketOptions = Object */

/**
 * @extends EventEmitter
 *
 * @emits Server#ready
 * @emits Server#error
 *
 * @borrows Client#encodeMessage as #encodeMessage
 */
class Server extends EventEmitter {
  /* ::
  authTimeout: number
  WebSocketServer: constructor
  clients: Map<string, Object>
  connectionHook: ConnectionHook
  socketOptions: Object
  wss: Object
   */

  /**
   * Starts a server.
   *
   * @param {Object} wssOptions Options that are passed to ws server.
   * @param {Server.ServerOptions} [serverOptions] Server options.
   * @param {Client.SocketOptions} [socketOptions] Socket options.
   */
  constructor (wssOptions /* : Object */,
    serverOptions /* : ServerOptions */ = {},
    socketOptions /* : SocketOptions  */ = {}) {
    super()
    this.socketOptions = { WebSocket }
    assign(this.socketOptions, socketOptions)
    assign(this, defaults, serverOptions)
    this.clients = new Map()
    this._setEvents(wssOptions)
  }

  _setEvents (wssOptions /* : Object */) /* : void */ {
    /**
     * Emits a ready event.
     * @private
     * @event Server#ready
     */
    this.wss = new this.WebSocketServer(wssOptions, (error) => {
      error ? this.emit('error', error) : this.emit('ready')
    })
    /**
     * Emits wss error events. Does not throw if there are no
     * listeners.
     * @private
     * @event Server#error
     * @param {Error} error Error.
     */
    this.wss.on('error', this.emit.bind(this))
    this.wss.on('connection', socket => this._onConnection(socket))
  }

  _onConnection (socket /* : Object & EventEmitter */) /* : void */ {
    const timeout = setTimeout(
      socket.close.bind(socket, CLOSE_FORBIDDEN), this.authTimeout)
    socket.once('message', data => this._addClient(socket, data, timeout))
  }

  _addClient (socket /* : EventEmitter */,
    data /* : any */,
    timeout /* : number */) /* : void */ {
    let client
    clearTimeout(timeout)
    uid(18).then(id => {
      client = new Client(null, assign({ socket, id }, this.socketOptions))
      client.autoReconnect = false
      this.clients.set(client.id, client)
      client.on('close', () => this._removeClient(client.id))
      if (this.connectionHook) {
        return attempt(() => client.decoder(data))
          .then(authData => this.connectionHook(client, authData))
      }
    }).then(authReplyData => {
      if (client._isOpen()) {
        client._setEvents()
        client.connected = true
        client.send('connect', authReplyData).catch(() => {})
        client._ping()
      }
    }).catch(error => {
      /* istanbul ignore else */
      if (client) {
        const str = error.toString()
        client.close(CLOSE_FORBIDDEN, str)
      }
    })
  }

  _removeClient (id /* : string */) /* : void */ {
    this.clients.delete(id)
  }

  /**
   * Git a client by id.
   * @param {string} id Client id.
   * @returns {Client|undefined} Client if found.
   */
  getClient (id /* : string */) /* : Client */ {
    return this.clients.get(id)
  }

  encodeMessage (event /* : string */, ...args /* : Array<any> */) /* : any */ {
    const encoder = this.socketOptions.encoder || JSON.stringify
    return attempt(() => encoder({ name: event, args }))
  }

  /**
   * Closes a server.
   * @param {code} [code=1000] Code as per WebSocket spec.
   * @returns {Promise<undefined>} Promise.
   */
  close (code /* : number */ = 1000) /* : Promise<void> */ {
    for (const [, client] of this.clients) {
      client.close(code)
    }
    this.clients.clear()
    return fromCallback(cb => this.wss.close(cb))
  }
}

module.exports = Server


src/Client.js
'use strict'
/* global WebSocket */

// Shared code for node and clients/browsers

const EventEmitter = require('eventemitter3')
const { assign, attempt, fromCallback, hasOwnProperty } = require('./utils')

const blacklist = ['close', 'open', 'error', 'pong', 'retry']

// utils

const concat = [].concat.bind([])

const emit = EventEmitter.prototype.emit

function isBlacklistedEvent (name) {
  return blacklist.indexOf(name) >= 0
}

function getOwnProp (obj, name) {
  return hasOwnProperty(obj, name) ? obj[name] : undefined
}

// errors

/**
 * {@link Client#send}/{@link Client#invoke} is rejected with this
 * error when connection is closed before a message is send or reply
 * received (for `invoke` only).
 *
 * @class
 * @augments Error
 * @memberof Client
 * @param {number} [id] Message id.
 */
function ConnectionError (id) {
  this.name = 'ConnectionError'
  this.id = id
}

ConnectionError.prototype = Object.create(Error.prototype)

/**
 * {@link Client#invoke} is rejected with this error when no reply is
 * received before {@link Client.SocketOptions} `ackTimeout`.
 *
 * @class
 * @augments Error
 * @memberof Client
 * @param {number} id Message id.
 */
function TimeoutError (id) {
  this.name = 'TimeoutError'
  this.id = id
}

TimeoutError.prototype = Object.create(Error.prototype)

/**
 * {@link Client#invoke} is rejected with this error by the other side
 * when no handler is found for a procedure.
 *
 * @class
 * @augments Error
 * @memberof Client
 * @param {string} procedure Name.
 */
function NoProcedureError (procedure) {
  this.name = 'NoProcedureError'
  this.procedure = procedure
}

NoProcedureError.prototype = Object.create(Error.prototype)

// validation

function validateId (id) {
  return typeof id === 'number' && (id % 1) === 0 && id > 0
}

function validate (message) {
  let passed = false
  if (message instanceof Object) {
    if (message.name) { // message
      if (typeof message.name === 'string' && message.args instanceof Array) {
        const nprops = Object.keys(message).length
        if (nprops === 2 || (nprops === 3 && validateId(message.id))) {
          passed = true
        }
      }
    } else { // ack message
      const nprops = Object.keys(message).length
      if (nprops === 2 && validateId(message.id)) {
        if (hasOwnProperty(message, 'error') || hasOwnProperty(message, 'result')) {
          passed = true
        }
      }
    }
  }
  if (!passed) {
    throw new Error('Validation error')
  }
}

// ack

class Ack {
  constructor (id, timeout, cb) {
    this.id = id
    this.promise = new Promise((resolve, reject) => {
      this.resolve = resolve
      this.reject = reject
    })
    this.cb = cb
    this.timeout = setTimeout(() => this.forceNack(new TimeoutError(id)), timeout)
  }

  forceNack (error) {
    clearTimeout(this.timeout)
    this.cb()
    this.reject(error)
  }

  settle (message) {
    clearTimeout(this.timeout)
    this.cb()
    if (hasOwnProperty(message, 'error')) {
      this.reject(message.error)
    } else {
      this.resolve(message.result)
    }
  }
}

// client

/**
 * General format for all data that is sent or received over a
 * websocket.
 *
 * @typedef {Object} Client.Message
 *
 * @property {number} [id]
 * @property {string} [name]
 * @property {Array} [args]
 * @property {Object} [result]
 * @property {Object} [error]
 */

/**
 * Messages encoder. May also return promises for an asynchronous
 * execution.
 *
 * @callback Client.Encoder
 * @param {Client.Message} message Message.
 * @return {Promise<Object>|Object} Data to send.
 */

/**
 * Messages decoder. May also return promises for an asynchronous
 * execution.
 *
 * @callback Client.Decoder
 * @param {Object} data Received data.
 * @return {Promise<Client.Message>|Client.Message} Message.
 */

/**
 * Receive hook is run when a client receives a valid message via a
 * websocket. May also return promises for an asynchronous execution.
 *
 * @callback Client.ReceiveHook
 * @param {Client.Message} message Message.
 * @return {Promise<undefined>|undefined} Promise, if it is rejected no
 * handlers will be called.
 */

/**
 * Send hook is run when a client sends any message via a
 * websocket. May also return promises for an asynchronous execution.
 *
 * @callback Client.SendHook
 * @param {Client.Message|Object} message Message or object if
 * `isEncoded` is `true`.
 * @param {boolean} isEncoded If a message has been already encoded
 * via {@link Client#encodeMessage} or {@link Server#encodeMessage}.
 * @return {Promise<undefined>|undefined} Promise, if it is rejected no
 * handlers will be called.
 */

/**
 * @typedef {Object} Client.RetryConfig
 *
 * @property {number} [factor=2]
 * @property {number} [maxTimeout=Infinity]
 * @property {number} [minTimeout=1000]
 * @property {boolean} [randomize=true]
 * @property {number} [retries=10]
 */

/**
 * @typedef {Object} Client.SocketOptions
 *
 * @property {number} [ackTimeout=20000] Result wait timeout for
 * {@link Client#invoke} in ms.
 * @property {Object} [auth={}] Auth data.
 * @property {boolean} [autoReconnect=true] Enable auto reconnect.
 * @property {Client.RetryConfig} [autoReconnectOptions] Auto
 * reconnect config.
 * @property {string} [binaryType='arraybuffer'] W3C WebSocket
 * binary data type.
 * @property {Client.Decoder} [decoder=JSON.parse] Messages decoder.
 * @property {Client.Encoder} [encoder=JSON.stringify] Messages
 * encoder.
 * @property {function} [errorFormatter=String] Converter for JS
 * errors to some network format.
 * @property {number} [pingInterval=20000] Ping interval in ms.
 * @property {number} [pingTimeout=20000] Ping timeout in ms.
 * @property {string|Array<string>} [protocols='ws-messaging']
 * WebSocket protocols.
 * @property {Client.ReceiveHook} [receiveHook] Receive hook.
 * @property {Client.SendHook} [rendHook] Send hook.
 * @property {boolean} [skipValidation=false] Skips build-in
 * messages validation.
 * @property {Object} [WebSocket=undefined] Alternative websocket
 * constructor, if it is undefined then a global WebSocket is used.
 * @property {boolean} [w3c=undefined] If WebSocket is using a w3c
 * send API, or a ws one (from Node.js server implementation with a
 * callback). By default if a global value is used, then it is `true`
 * and `false` otherwise.
 * @property {Object} [wsOptions] Additional options to pass to ws
 * socket constructor.
 */

const retryConfig = {
  factor: 2,
  maxTimeout: Infinity,
  minTimeout: 1000,
  randomize: true,
  retries: 10
}

const defaults = {
  ackTimeout: 20000,
  auth: {},
  autoReconnect: true,
  autoReconnectOptions: retryConfig,
  binaryType: 'arraybuffer',
  decoder: JSON.parse,
  encoder: JSON.stringify,
  errorFormatter: String,
  pingInterval: 20000,
  pingTimeout: 20000,
  protocols: 'ws-messaging',
  receiveHook: null,
  sendHook: null,
  skipValidation: false,
  WebSocket: undefined,
  w3c: undefined,
  wsOptions: undefined
}

/**
 * @extends EventEmitter
 *
 * @emits Client#close
 * @emits Client#open
 * @emits Client#error
 * @emits Client#connect
 */
class Client extends EventEmitter {
  /**
   * Creates a client.
   *
   * @param {string} url WebSocket connection url.
   * @param {Client.SocketOptions} [options] Socket options.
   */
  constructor (url, options = {}) {
    super()
    this.url = url
    /**
     * Client id. Server-side only.
     * @member {number} id
     * @memberof Client
     * @instance
     * @readonly
     */
    assign(this, defaults, options)
    this.retryConfig = {}
    assign(this.retryConfig, retryConfig, options.autoReconnectOptions)
    if (!this.WebSocket) {
      this.WebSocket = WebSocket
      this.w3c = this.w3c === undefined ? true : this.w3c
    }
    /**
     * If true, then a client is connected.
     * @member {boolean}
     * @readonly
     */
    this.connected = false
    this.counter = 1
    this.data = {}
    this.handlers = {}
    this.pendingAcks = {}
    this.attempt = 0
    /**
     * If true, then a client was closed via a close method or an auth
     * error occurred.
     * @member {boolean}
     * @readonly
     */
    this.terminated = false
    this.register('ping', () => Promise.resolve())
    this.reconnect()
  }

  _onMessage (data) {
    let message
    attempt(() => this.decoder(data.data))
      .then(msg => { message = msg })
      .then(() => { if (!this.skipValidation) { validate(message) } })
      .then(() => { if (this.receiveHook) { return this.receiveHook(message) } })
      .then(() => this._dispatch(message))
    /**
     * Emitted when the other side failed to decode or validate a
     * websocket message, namely an error is occurred inside either
     * `decoder` or `receiveHook`.
     * @event Client#preprocessingError
     * @param {Object} error Converted error.
     */
      .catch(error => this.send('preprocessingError', this.errorFormatter(error)))
  }

  _setEvents () {
    /**
     * Emits w3c onopen WebSocket events.
     * @event Client#open
     */
    this.socket.onopen = emit.bind(this, 'open')
    /**
     * Emits w3c onerror WebSocket events. Does not throw if there are
     * no listeners.
     * @event Client#error
     * @param {Error} error Error.
     */
    this.socket.onerror = emit.bind(this, 'error')
    this.socket.onclose = this._onClose.bind(this)
    this.socket.onmessage = this._onMessage.bind(this)
  }

  _ping () {
    this.pingTimeoutId = setTimeout(() => {
      emit.call(this, 'ping')
      const timeout = this.pingTimeout
      const { message, promise } = this._makeMessage('ping', [], true, timeout)
      this._send(message).then(() => promise)
        .then(() => emit.call(this, 'pong'))
        .then(() => this._ping())
        .catch(() => this.close(4008, 'Ping timeout', false))
    }, this.pingInterval)
  }

  _isOpen () {
    return this.socket &&
      (this.socket.readyState === 0 || this.socket.readyState === 1)
  }

  _reconnect () {
    const { factor, maxTimeout, minTimeout, randomize, retries } = this.retryConfig
    if (this.attempt >= retries || this.terminated) { return }
    const rand = 1 + (randomize ? Math.random() : 0)
    const timeout = Math.min(rand * minTimeout * Math.pow(factor, this.attempt), maxTimeout)
    this.reconnectTimeoutId = setTimeout(this.reconnect.bind(this), timeout)
    this.attempt++
  }

  _open () {
    clearTimeout(this.reconnectTimeoutId)
    /**
     * Underlying websocket.
     * @member {WebSocket}
     * @readonly
     */
    this.socket = new this.WebSocket(this.url, this.protocols, this.wsOptions)
    if (this.w3c) { this.socket.binaryType = this.binaryType }
    this.connectHandler = () => {
      this.connected = true
      this.attempt = 0
      clearTimeout(this.authTimeoutId)
      this._ping()
    }
    this.openHandler = () => {
      this._send(this.auth, { isAuth: true })
      this.authTimeoutId = setTimeout(
        this.close.bind(this, 4008, 'Auth timeout', false),
        this.ackTimeout)
      this.once('connect', this.connectHandler)
    }
    this.once('open', this.openHandler)
    this._setEvents()
    if (this.attempt > 0) {
      /**
       * Emits retry events when auto reconnecting.
       * @event Client#retry
       * @param {number} attempt Attempt number starting from `1`.
       */
      emit.call(this, 'retry', this.attempt)
    }
  }

  _onClose (ev) {
    this.connected = false
    clearTimeout(this.pingTimeoutId)
    clearTimeout(this.authTimeoutId)
    clearTimeout(this.reconnectTimeoutId)
    this.off('connect', this.connectHandler)
    this.off('open', this.openHandler)
    if (ev.code === 4003 || !this.url) { this.terminated = true }
    for (const id in this.pendingAcks) {
      /* istanbul ignore else */
      if (hasOwnProperty(this.pendingAcks, id)) {
        const ack = this.pendingAcks[id]
        ack.forceNack(new ConnectionError(id))
      }
    }
    this.pendingAcks = {}
    if (!this.terminated && this.autoReconnect) { this._reconnect() }
    /**
     * Emits w3c onclose WebSocket events.
     * @event Client#close
     * @param {CloseEvent} data Close event data.
     */
    emit.call(this, 'close', ev)
  }

  _makeMessage (name, args, needsAck, ackTimeout = this.ackTimeout) {
    let promise
    const message = { name, args }
    if (needsAck) {
      const id = this.counter++
      const ack = new Ack(id, ackTimeout, () => delete this.pendingAcks[id])
      this.pendingAcks[id] = ack
      promise = ack.promise
      message.id = id
    }
    return { message, promise }
  }

  /**
   * Socket connection is open and client has passed an auth
   * check. Client-side only.
   * @event Client#connect
   * @param {Object|undefined} data Auth reply data.
   */
  _dispatch (message) {
    if (message.name) {
      if (isBlacklistedEvent(message.name)) { return }
      if (message.name === 'connect' && this.connected) { return }
      if (message.id) {
        const id = message.id
        const fn = getOwnProp(this.handlers, message.name)
        if (fn) {
          attempt(() => fn.apply(null, message.args))
            .then((result = null) => this._send({ id, result }))
            .catch(error => this._send({ id, error: this.errorFormatter(error) }))
        } else {
          const error = this.errorFormatter(new NoProcedureError(message.name))
          this._send({ id, error })
        }
      } else {
        emit.apply(this, concat(message.name, message.args))
      }
    } else {
      const ack = this.pendingAcks[message.id]
      if (ack) { ack.settle(message) }
    }
  }

  _send (message, { skipEncoder = false, isAuth = false } = {}) {
    return attempt(() => this.sendHook ? this.sendHook(message, skipEncoder) : null)
      .then(() => skipEncoder ? message : this.encoder(message))
      .then(data => {
        if (!this.connected && !isAuth) {
          const id = skipEncoder ? undefined : message.id
          throw new ConnectionError(id)
        }
        if (this.w3c) {
          return this.socket.send(data)
        } else {
          const binary = typeof data !== 'string'
          return fromCallback(cb => this.socket.send(data, { binary }, cb))
        }
      })
  }

  /**
   * Send an event, no reply. Use {@link on} or {@link once} methods
   * to listen events on a recipient side. Reserved event names
   * (__MUST NOT__ be used): `connect`, `close`, `open`, `error`,
   * `ping`, `pong`, `retry`.
   * @param {string} event Event name.
   * @param {*} [args] Arguments.
   * @returns {Promise<undefined>} Resolves when a data has been sent.
   */
  send (event, ...args) {
    const { message } = this._makeMessage(event, args, false)
    return this._send(message)
  }

  /**
   * Send a message encoded by {@link Client#encodeMessage} or {@link
   * Server#encodeMessage}, useful for identical messages
   * broadcasting.
   * @param {Object} data Result of {@link Client#encodeMessage}.
   * @returns {Promise<undefined>} Resolves when a data has been sent.
   */
  sendEncoded (data) {
    return this._send(data, { skipEncoder: true })
  }

  /**
   * Encode a message for a later use with {@link Client#sendEncoded}.
   * Reserved event names (__MUST NOT__ be used): `connect`, `close`,
   * `open`, `error`, `ping`, `pong`, `retry`.
   * @param {string} event Event name.
   * @param {*} [args] Arguments.
   * @returns {Object} Encoded message.
   */
  encodeMessage (event, ...args) {
    const { message } = this._makeMessage(event, args, false)
    return attempt(() => this.encoder(message))
  }

  /**
   * Invoke an RPC procedure. Use {@link Client#register} method to
   * assign an RPC method handler. Reserved procedure names (__MUST
   * NOT__ be used): `connect`, `close`, `open`, `error`, `ping`,
   * `pong`, `retry`.
   * @param {string} name Procedure name.
   * @param {*} [args] Arguments.
   * @returns {Promise<Object>} Resolves or rejects when a reply is
   * received.
   */
  invoke (name, ...args) {
    const { message, promise } = this._makeMessage(name, args, true)
    return this._send(message).then(() => promise)
  }

  /**
   * Register an RPC handler. Each name must have no more than a one
   * handler, so it throws an error on a duplicate handler
   * registration attempt. Use {@link Client#invoke} to call a method.
   * @param {string} name Procedure name.
   * @param {function} handler A function that returns a Promise.
   */
  register (name, handler) {
    if (getOwnProp(this.handlers, name)) {
      throw new Error(`Can't register a duplicate RPC handler for ${name}`)
    }
    this.handlers[name] = handler
  }

  /**
   * Reconnect. Client-side only.
   */
  reconnect () {
    this.terminated = false
    if (this._isOpen()) {
      // eslint-disable-next-line no-useless-return
      return
    } else if (this.WebSocket && this.url) {
      this._open()
    } else {
      throw new Error('Malformed configuration options')
    }
  }

  /**
   * Closes a client connection.
   * @param {number} [code=1000] Code as per WebSocket spec.
   * @param {string} [str] Optional string.
   * @param {boolean} [terminate=true] Disable reconnect.
   */
  close (code = 1000, str, terminate = true) {
    if (!this.terminated) {
      this.terminated = terminate
      if (this._isOpen()) { this.socket.close(code, str) }
    }
  }
}

/**
 * Alias for {@link Client#send}.
 * @method
 * @name Client#emit
 */
Client.prototype.emit = Client.prototype.send

/**
 * Alias for {@link Client#sendEncoded}.
 * @method
 * @name Client#emitEncoded
 */
Client.prototype.emitEncoded = Client.prototype.sendEncoded

Client.ConnectionError = ConnectionError
Client.NoProcedureError = NoProcedureError
Client.TimeoutError = TimeoutError

module.exports = Client
